### PR TITLE
Flood: allow to remove downloads when seeding criteria have been met

### PR DESCRIPTION
**Pull request by jesec**
This change implements the new DownloadSeedConfigProvider for Flood.

See 411be4d0116f0739bb9c71235312d0c5a26dd3a2 and #3632.

### CODE REVIEW

**Comment by jesec** [comment]
cc: @Taloth

**Comment by leaty** [comment]
I just wanted to say I'm a big fan of Flood. It's hands down the best torrent webUI, big props to you. One question in regards to the PR, Is there a way for Flood to also support `SeedTime`?

**Comment by jesec** [comment]
> I just wanted to say I'm a big fan of Flood. It's hands down the best torrent webUI, big props to you. One question in regards to the PR, Is there a way for Flood to also support `SeedTime`?

Flood API doesn't expose download finished time yet, so it is not possible at the moment.

File: src/NzbDrone.Core/Download/Clients/Flood/Flood.cs [hunk 1]
```diff
@@ -143,7 +148,16 @@ public override IEnumerable<DownloadClientItem> GetItems()
                     item.Status = DownloadItemStatus.Paused;
                 }
 
-                item.CanMoveFiles = item.CanBeRemoved = false;
+                if (item.Status == DownloadItemStatus.Completed)
+                {
+                    // Grab cached seedConfig
+                    var seedConfig = _downloadSeedConfigProvider.GetSeedConfiguration(item.DownloadId);
+
+                    if (item.SeedRatio >= seedConfig.Ratio)
```
Comment (Taloth) [inline src/NzbDrone.Core/Download/Clients/Flood/Flood.cs:156]: check `seedConfig != null &&`

File: src/NzbDrone.Core/Download/Clients/Flood/Flood.cs [hunk 2]
```diff
@@ -18,17 +18,20 @@ namespace NzbDrone.Core.Download.Clients.Flood
...
         private readonly IFloodProxy _proxy;
+        private readonly IDownloadSeedConfigProvider _downloadSeedConfigProvider;
 
         public Flood(IFloodProxy proxy,
                         ITorrentFileInfoReader torrentFileInfoReader,
                         IHttpClient httpClient,
                         IConfigService configService,
                         IDiskProvider diskProvider,
                         IRemotePathMappingService remotePathMappingService,
+                        IDownloadSeedConfigProvider downloadSeedConfigProvider,
```
Comment (Taloth) [inline src/NzbDrone.Core/Download/Clients/Flood/Flood.cs:29]: Nitpick:  Move this to above ITorrentFileInfoReader, that way the last parameters all go to the base class.

File: src/NzbDrone.Core/Download/Clients/Flood/Flood.cs (hunk 1 above)
Comment (Taloth) [reply src/NzbDrone.Core/Download/Clients/Flood/Flood.cs:156]: Also add comment that SeedTime isn't supported. 
Maybe add that to the Message too?

**Review by Taloth — CHANGES_REQUESTED** [review]
Some nits. 

I kinda forgot Flood in the migration, you might want to change `Implementation = 'RTorrent'` to `Implementation IN ('RTorrent', 'Flood')` in the existing db migration, so that Remove gets disabled (for backward compat) when ppl are (auto-)updating to a new sonarr version.

**Comment by jesec** [comment]
> Some nits.
> 
> I kinda forgot Flood in the migration, you might want to change `Implementation = 'RTorrent'` to `Implementation IN ('RTorrent', 'Flood')` in the existing db migration, so that Remove gets disabled (for backward compat) when ppl are (auto-)updating to a new sonarr version.

Done.

File: src/NzbDrone.Core/Download/Clients/Flood/Flood.cs (hunk 2 above)
Comment (jesec) [reply src/NzbDrone.Core/Download/Clients/Flood/Flood.cs]: Done.

File: src/NzbDrone.Core/Download/Clients/Flood/Flood.cs (hunk 1 above)
Comment (jesec) [reply src/NzbDrone.Core/Download/Clients/Flood/Flood.cs]: A `dateFinished` field is going to be added in newer versions of Flood, so I revised the patch to support that.

**Comment by Taloth** [comment]
Thank you.

**Pull request merged**